### PR TITLE
    Add --noclobber support to safe

### PR DESCRIPTION
**Pull request by geofffranks**
`safe --noclobber` will throw a message up to the user
    that they tried to overwrite an existing credential.
    This is supported in all known write-causing commands,
    except `safe import`.

    When existing credentials are encountered, safe will exit 0,
    as it successfully avoided clobbering the credential.

    If `--quiet` is specified, warning messages for clobbering-noops will be suppressed

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,4 @@ release
 ci/credentials.yml
 /t
 /vaults
+*.swp
```
Comment (jhunt) [inline .gitignore]: This should go in your `~/.gitignore` 😉

File: main.go [hunk 2]
```diff
@@ -618,9 +626,9 @@ sense when you are pasting in credentials from an external password manager
 like 1password or Lastpass.
 `,
 	}, func(command string, args ...string) error {
-		//writeHelper is defined right about set, which is defined right about this
+		//writeHelper is defined right above the ask subcommand
```
Comment (jhunt) [inline main.go]: Do we _really_ need to use comments to identify where functions are defined?

File: main.go [hunk 3]
```diff
@@ -1059,7 +1093,14 @@ NBITS defaults to 2048.
 		if err != nil && !vault.IsNotFound(err) {
 			return err
 		}
-		if err = s.DHParam(bits); err != nil {
+		exists := (err == nil)
+		if opt.SkipIfExists && exists && s.Has("dhparam-pem") {
+			if !opt.Quiet {
+				ansi.Fprintf(os.Stderr, "@R{Cowardly refusing to generate a DH Param in} @C{%s} @R{as it is already present in Vault}\n", path)
```
Comment (jhunt) [inline main.go]: Can we spell out that `DH Param` is `Diffie-Helman key exchange parameters`?

File: main.go [hunk 4]
```diff
@@ -1070,6 +1111,10 @@ NBITS defaults to 2048.
 		Usage:   "safe echo Your Message Here:",
 		Type:    NonDestructiveCommand,
 	}, func(command string, args ...string) error {
+		// --no-clobber is ignored here, because there's no context of what you're
+		// about to be writing after a prompt, so not sure if we should our shouldn't prompt
```
Comment (jhunt) [inline main.go]: or

File: main.go [hunk 5]
```diff
@@ -1021,7 +1048,14 @@ be PEM-encoded.
 			if err != nil && !vault.IsNotFound(err) {
 				return err
 			}
-			if err = s.RSAKey(bits); err != nil {
+			exists := (err == nil)
+			if opt.SkipIfExists && exists && (s.Has("private") || s.Has("public")) {
```
Comment (jhunt) [inline main.go]: Do RSA keypair not have fingerprints?

File: vault/vault.go [hunk 6]
```diff
@@ -467,35 +466,52 @@ func (v *Vault) deleteIfPresent(path string) error {
...
 // Returns KeyNotFoundError if there is no such specified key in the secret at oldpath
-func (v *Vault) Copy(oldpath, newpath string) error {
+func (v *Vault) Copy(oldpath, newpath string, skipIfExists bool, quiet bool) error {
 	if err := v.verifySecretExists(oldpath); err != nil {
 		return err
 	}
+	if skipIfExists {
+		if _, err := v.Read(newpath); err == nil {
+			if !quiet {
+				ansi.Fprintf(os.Stderr, "@R{Cowardly refusing to copy/move data into} @C{%s}@R{, as it would clobber existing data}\n", newpath)
```
Comment (jhunt) [inline vault/vault.go]: Can we reword this to "as that would" or "as doing so would"?

**Comment by dennisjbell** [comment]
Isn't it convention that negative Boolean options are in the form of --no-<whatever>?  From the Principle of Least Surprise, I'd recommend we use --no-clobber unless there's a compelling reason we shouldn't.

File: main.go (hunk 5 above)
Comment (geofffranks) [inline main.go]: Addressed

**Comment by geofffranks** [comment]
Switched to --no-clobber

**Comment by geofffranks** [comment]
@jhunt all feedback addressed + @dennisjbell 's as well
